### ISSUE TITLE
Building on a clean system

### ISSUE THREAD

**Issue by maxsvetlik** [ut-amrl/ut_multirobot_sim#22]
I'm trying to build this project on a clean system. Unless I'm missing something, the README is missing steps to handle submodules and/or I'm manually adding submodules incorrectly.

My steps include:
1. git clone git@github.com:ut-amrl/f1tenth_simulator.git
2. cd f1tenth_simulator/submodules
3. git submodule init
4. git submodule update
5. cd ../ && make

After step 4, the submodules are successfully cloned into their respective folders. However, after running `make` I get the following errors:

```
[ 44%] Building CXX object CMakeFiles/simulator.dir/src/simulator/simulator_main.cpp.o
In file included from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/entity_base.h:28:0,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/vector_map.h:27,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.h:44,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator_main.cpp:30:
/home/max/Documents/programovani/f1tenth_simulator/submodules/shared/math/poses_2d.h:26:36: fatal error: shared/common_includes.h: No such file or directory
compilation terminated.
```
`common_includes.h` is indeed not in the root of the `shared/` directory. 

To those that have built the project successfully, what are the deltas in our setups?

**Comment by jarrettHoltz** [comment]
So there are a couple of issues with the build of master as it currently stands. This one is actually an issue with that submodule trying to include a file that doesn't exist that I think people have been fixing locally by removing that include from poses_2d.h.

There are two things in progress to fix this. 
1. I'm about to merge in PR to master that should allow this to build in g++ at least. (There is another issue to resolve with Clang.)

2. Setup a Travis CI build for this so these types of things are monitored better.

**Comment by jarrettHoltz** [comment]
You should be able to build on a clean system with g++ now.

**Comment by maxsvetlik** [comment]
After installing the lua dependencies from `config_reader` along with pulling the recent changes to master I do get further in the build process. My build errors out before the Eigan issue mentioned in #24. Build output follows

```
[ 55%] Building CXX object CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o
In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
                 from /usr/include/c++/5/bits/allocator.h:46,
                 from /usr/include/c++/5/memory:63,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
/usr/include/c++/5/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’:
/usr/include/c++/5/bits/alloc_traits.h:530:4:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >]’
/usr/include/c++/5/bits/hashtable_policy.h:1955:37:   required from ‘std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true>]’
/usr/include/c++/5/bits/hashtable_policy.h:180:58:   required from ‘std::__detail::_AllocNode<_NodeAlloc>::__node_type* std::__detail::_AllocNode<_NodeAlloc>::operator()(_Arg&&) const [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >&; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> >; std::__detail::_AllocNode<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true>]’
/usr/include/c++/5/bits/hashtable.h:1703:18:   required from ‘std::pair<typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_insert(_Arg&&, const _NodeGenerator&, std::true_type) [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >&; _NodeGenerator = std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> > >; _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>; std::true_type = std::integral_constant<bool, true>]’
/usr/include/c++/5/bits/hashtable_policy.h:715:55:   required from ‘std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type = std::pair<std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>, bool>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::value_type = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’
/usr/include/c++/5/bits/unordered_map.h:433:31:   required from ‘std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator, bool> std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Tp = std::unique_ptr<config_reader::config_types::TypeInterface>; _Hash = std::hash<std::__cxx11::basic_string<char> >; _Pred = std::equal_to<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’
/home/max/Documents/programovani/f1tenth_simulator/submodules/config_reader/include/config_reader/macros.h:94:31:   required from ‘const CPPType& config_reader::InitVar(const string&) [with CPPType = std::__cxx11::basic_string<char>; ConfigType = config_reader::config_types::ConfigString; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:60:1:   required from here
/usr/include/c++/5/ext/new_allocator.h:120:4: error: use of deleted function ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = std::unique_ptr<config_reader::config_types::TypeInterface>]’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^
In file included from /usr/include/c++/5/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/5/memory:62,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
/usr/include/c++/5/bits/stl_pair.h:127:17: note: ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = std::unique_ptr<config_reader::config_types::TypeInterface>]’ is implicitly deleted because the default definition would be ill-formed:
       constexpr pair(const pair&) = default;
                 ^
/usr/include/c++/5/bits/stl_pair.h:127:17: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = config_reader::config_types::TypeInterface; _Dp = std::default_delete<config_reader::config_types::TypeInterface>]’
In file included from /usr/include/c++/5/memory:81:0,
                 from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
/usr/include/c++/5/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
CMakeFiles/simulator.dir/build.make:192: recipe for target 'CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o' failed
make[3]: *** [CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o] Error 1
CMakeFiles/Makefile2:68: recipe for target 'CMakeFiles/simulator.dir/all' failed
make[2]: *** [CMakeFiles/simulator.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make[1]: *** [all] Error 2
Makefile:10: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by jarrettHoltz** [comment]
Can you post your compiler and Ubuntu versions?
Config reader has an issue I haven't looked deeply into yet when you switch to building in Xenial (in the CI system). I think this is related.

**Comment by maxsvetlik** [comment]
g++ version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Linux kernel: 4.15.0-96-generic
Ubuntu: 16.04.6 Xenial

I actually was thinking I was in 18.04. I may switch over to 18.04 and see if I can build there.

**Comment by jarrettHoltz** [comment]
If the problem is some difference between Xenial and Bionic it's something we need to try and fix anyway. Thought it would be good to identify that as the problem.

**Comment by maxsvetlik** [comment]
Indeed, the simulator compiles successfully on a freshly installed 18.04 with g++ version 7.5.0.

Another note: the RViz file config file assumes that you have (rviz_mouse_pub)[https://github.com/ut-amrl/rviz_mouse_pub] in your ROS_PACKAGE_PATH. We might want to add that as a submodule

**Comment by YuqianJiang** [comment]
> After installing the lua dependencies from `config_reader` along with pulling the recent changes to master I do get further in the build process. My build errors out before the Eigan issue mentioned in #24. Build output follows
> 
> ```
> [ 55%] Building CXX object CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o
> In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
>                  from /usr/include/c++/5/bits/allocator.h:46,
>                  from /usr/include/c++/5/memory:63,
>                  from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
> /usr/include/c++/5/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’:
> /usr/include/c++/5/bits/alloc_traits.h:530:4:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >]’
> /usr/include/c++/5/bits/hashtable_policy.h:1955:37:   required from ‘std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<config_reader::config_types::TypeInterface, std::default_delete<config_reader::config_types::TypeInterface> > >&}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true>]’
> /usr/include/c++/5/bits/hashtable_policy.h:180:58:   required from ‘std::__detail::_AllocNode<_NodeAlloc>::__node_type* std::__detail::_AllocNode<_NodeAlloc>::operator()(_Arg&&) const [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >&; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> >; std::__detail::_AllocNode<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true>]’
> /usr/include/c++/5/bits/hashtable.h:1703:18:   required from ‘std::pair<typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_insert(_Arg&&, const _NodeGenerator&, std::true_type) [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >&; _NodeGenerator = std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, true> > >; _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>; std::true_type = std::integral_constant<bool, true>]’
> /usr/include/c++/5/bits/hashtable_policy.h:715:55:   required from ‘std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type = std::pair<std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>, bool>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::value_type = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’
> /usr/include/c++/5/bits/unordered_map.h:433:31:   required from ‘std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator, bool> std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Tp = std::unique_ptr<config_reader::config_types::TypeInterface>; _Hash = std::hash<std::__cxx11::basic_string<char> >; _Pred = std::equal_to<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> > >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >, false, true>; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<config_reader::config_types::TypeInterface> >]’
> /home/max/Documents/programovani/f1tenth_simulator/submodules/config_reader/include/config_reader/macros.h:94:31:   required from ‘const CPPType& config_reader::InitVar(const string&) [with CPPType = std::__cxx11::basic_string<char>; ConfigType = config_reader::config_types::ConfigString; std::__cxx11::string = std::__cxx11::basic_string<char>]’
> /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:60:1:   required from here
> /usr/include/c++/5/ext/new_allocator.h:120:4: error: use of deleted function ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = std::unique_ptr<config_reader::config_types::TypeInterface>]’
>   { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
>     ^
> In file included from /usr/include/c++/5/bits/stl_algobase.h:64:0,
>                  from /usr/include/c++/5/memory:62,
>                  from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
> /usr/include/c++/5/bits/stl_pair.h:127:17: note: ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = std::unique_ptr<config_reader::config_types::TypeInterface>]’ is implicitly deleted because the default definition would be ill-formed:
>        constexpr pair(const pair&) = default;
>                  ^
> /usr/include/c++/5/bits/stl_pair.h:127:17: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = config_reader::config_types::TypeInterface; _Dp = std::default_delete<config_reader::config_types::TypeInterface>]’
> In file included from /usr/include/c++/5/memory:81:0,
>                  from /home/max/Documents/programovani/f1tenth_simulator/src/simulator/simulator.cpp:23:
> /usr/include/c++/5/bits/unique_ptr.h:356:7: note: declared here
>        unique_ptr(const unique_ptr&) = delete;
>        ^
> CMakeFiles/simulator.dir/build.make:192: recipe for target 'CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o' failed
> make[3]: *** [CMakeFiles/simulator.dir/src/simulator/simulator.cpp.o] Error 1
> CMakeFiles/Makefile2:68: recipe for target 'CMakeFiles/simulator.dir/all' failed
> make[2]: *** [CMakeFiles/simulator.dir/all] Error 2
> Makefile:127: recipe for target 'all' failed
> make[1]: *** [all] Error 2
> Makefile:10: recipe for target 'all' failed
> make: *** [all] Error 2
> ```

I'm having the same error on 16.04

**Comment by jarrettHoltz** [comment]
The simulator should now build in 16.04 as it is successfully building in 16.04 in the CI system.

**Comment by maxsvetlik** [comment]
Affirmative, I am able to build successfully after pulling updates from the submodules and updating from master. Good work everybody!
